### ISSUE TITLE
Regarding the learnable class token

### ISSUE THREAD

**Issue by moha23** [google-research/vision_transformer#83]
Hi, I'm trying to wrap my head around the use of 'cls' token. From the code https://github.com/google-research/vision_transformer/blob/4317e064a0a54b825b5b9ff634482954788b8d84/vit_jax/models.py#L251 what I understand is that if classification token is to be used, first a read-only variable 'cls' of shape (1, 1, c) is defined, initialised with zeros and repeated along first axis to get shape (n,1,c) and finally concatenated to patch embeddings x. A few, possibly naive, doubts:

1. So, do the values of 'cls' token keep getting updated as training proceeds? Or is it fixed to be 0 valued?
2.  When you say 'learnable' embedding, does it mean it is a trainable variable and the network learns in the training process what its values should be? 
3. And hence while testing this embedding will have some pre-trained values? But then how will this pre-trained value be different for different inputs from different classes?

**Comment by andsteing** [comment]
Hi

1. The `cls` token is defined via `self.param()` so it will be optimized by the optimizer exactly the same as the other model weights (e.g. dense kernels).
2. That's correct.
3. After training the embedding is fixed and it will be exactly the same for all inputs *in the first layer* - but due to interactions with the other tokens in every layer, the value will be input-dependent (and strongly correlate with the input's class) at the output layer.

**Comment by moha23** [comment]
Thanks a lot for clarifying!
